pkg/applications/installer.go

```diff
@@ -34,3 +39,79 @@ type ApplicationInstaller interface {
 	// Delete function uninstalls the application on the user-cluster and returns an error if the uninstallation has failed; this is idempotent.
 	Delete(ctx context.Context, log *zap.SugaredLogger, seedClient ctrlruntimeclient.Client, userClient ctrlruntimeclient.Client, applicationInstallation *appkubermaticv1.ApplicationInstallation) error
 }
+
+// ApplicationManager handles the installation / uninstallation of an Application on the user-cluster.
+type ApplicationManager struct {
+}
+
+// Apply creates the namespace where the application will be installed (if necessary) and installs the application.
+func (a *ApplicationManager) Apply(ctx context.Context, log *zap.SugaredLogger, seedClient ctrlruntimeclient.Client, userClient ctrlruntimeclient.Client, applicationInstallation *appkubermaticv1.ApplicationInstallation) error {
+	if err := a.applyNamespace(ctx, log, applicationInstallation, userClient); err != nil {
+		return err
+	}
+	// todo logic to download source and install app
+	return nil
+}
+
+// Delete uninstalls the application and deletes the namespace where the application was installed if necessary.
+func (a *ApplicationManager) Delete(ctx context.Context, log *zap.SugaredLogger, seedClient ctrlruntimeclient.Client, userClient ctrlruntimeclient.Client, applicationInstallation *appkubermaticv1.ApplicationInstallation) error {
+	// todo logic to uninstall application
+	return a.deleteNamespace(ctx, log, applicationInstallation, userClient)
+}
+
+// applyNamespace ensures namespace is created and has desired labels and annotations if applicationInstallation.Spec.Namespace.Create flag is set.
+func (a *ApplicationManager) applyNamespace(ctx context.Context, log *zap.SugaredLogger, applicationInstallation *appkubermaticv1.ApplicationInstallation, userClient ctrlruntimeclient.Client) error {
```

Comment:
I would rename this func to something like reconcileNamespace, because it contains the checking logic if the namespace should be created in the first place. And if I see it correctly, if "Create" is set to false, this func will not apply anything.

Alternatively the checking logic could also be moved into the `(*ApplicationManager) Apply`, but I do see how this makes the tests unnecessarily clunky.

Reply:
For the naming, I was hesitating, `reconcileNamespace`is good to me :)

> Alternatively the checking logic could also be moved into the (*ApplicationManager) Apply, but I do see how this makes the tests unnecessarily clunky.
 yes, moreover this function will also contain the logic to download the source and install the application with helm. So I think it's better to split logic into dedicated functions for more clarity and easy testing.